src/core/service.c

```diff
@@ -1440,6 +1440,119 @@ static bool service_exec_needs_notify_socket(Service *s, ExecFlags flags) {
         return s->notify_access != NOTIFY_NONE;
 }
 
+static int service_copy_exit_status_env(char **env, size_t *n_env, Service *s) {
+        assert(s);
+        assert(env);
+        assert(n_env);
+
+        /* Copy service exit metadata from 's' to env. */
+
+        int r = asprintf(env + *n_env, "SERVICE_RESULT=%s", service_result_to_string(s->result));
```

Comment:
we usually don't clobber immediately the caller's values, but accumulate in a local variable, and only assign to the return pointer at the end of the function when everything is good, so that on error the input is untouched and cannot be in a half-way state

Reply:
and the return pointer is usually a new empty pointer, and usually it's the caller that does the merging - but there are exceptions when required and the callee does the merging (especially when it's an array like in this case), but then it needs to be documented above the function, as it's an exception to the rule

Reply:
Thanks, I've corrected this in the latest diffs.